Autocomplete - Allow default display per entity

colemanw opened the pull request:
Overview
----------------------------------------
Allows admins to customize autocomplete dropdowns.

Before
----------------------------------------
Autocompletes use a default search display based on metadata.

After
----------------------------------------

Instead of always using an autogenerated default, a site admin can now choose a default display per-entity. E.g. a default display for autocompleting Contacts, Events, etc.

Technical Details
----------------------------------------
Setting is an array like `["Contact:MyContactDisplay", "Event:MyEventDisplay"]`

colemanw committed:
Autocomplete - Allow default display per entity

Allows admins to customize autocomplete dropdowns.

Instead of always using an autogenerated default,
a site admin can now choose a default display per-entity.
E.g. a default display for autocompleting Contacts, Events, etc.

eileenmcnaughton commented on Civi/Api4/Generic/AutocompleteAction.php:
```diff
@@ -132,6 +132,28 @@ public function _run(Result $result) {
+            ->setSelect(['*', 'type:name'])
+            ->addWhere('name', '=', $this->display)
+            ->addWhere('type', '=', 'autocomplete')
+            ->execute()->single();
```
I think you might be better to do `first()` & only continue if it has not been deleted without updating the setting

colemanw replied:
@eileenmcnaughton that's the idea with wrapping this in try/catch

colemanw committed:
Autocomplete - Add UI to set default

This adds a checkbox when configuring an autocomplete display in SearchKit,
which sets the display as the entity default.

eileenmcnaughton commented on settings/Search.setting.php:
```diff
@@ -220,6 +220,19 @@
+    'group_name' => 'Search Preferences',
+    'group' => 'Search Preferences',
+    'name' => 'autocomplete_displays',
+    'type' => 'Array',
```
Note if switching to a flat list then type is String & options look like  ` 'pseudoconstant' => ['callback' => 'CRM_Core_SelectValues::addressProvider'],`

colemanw committed:
Autocomplete - Add setting to Search Preferences screen

This includes validation when saving the setting to ensure only one
display per entity is selected, and that the search display exists
in the database.

eileenmcnaughton commented on CRM/Admin/Form/Setting/Search.php:
```diff
@@ -80,4 +80,80 @@ public static function enableOptionOne(&$value) {
+   *
+   * @return array
+   */
+  public static function getAutocompleteDisplays(): array {
```
I don't love these BAO-ish functions being on this form - but I assume there is precedent / no obvious other place

colemanw replied:
I don't love it either but it would be an easy refactor to move them to a better place if we think of one.

eileenmcnaughton commented on ext/search_kit/CRM/Search/BAO/SearchDisplay.php:
```diff
@@ -16,6 +16,38 @@
+    if (!$display->name || !$display->type || !$display->saved_search_id) {
+      $display->find(TRUE);
+    }
+    if ($display->type === 'autocomplete' && isset($params['is_autocomplete_default']) && (empty($params['check_permissions']) || CRM_Core_Permission::check('administer CiviCRM'))) {
```
I think I would use `'administer CiviCRM data'` here rather than `administer CiviCRM` (the former is more granular but implied by the latter)

colemanw replied:
@eileenmcnaughton it should match whatever permission is required by the **Settings - Search Preferences** form. I believe it's `administer CiviCRM` no?

eileenmcnaughton replied:
yeah - ok to leave
